### ISSUE TITLE
Python: run tests in separate derivation

### ISSUE THREAD

**Issue by FRidh** [NixOS/nixpkgs#26400]
## Issue description

Currently we run tests directly after a package has been build. Unfortunately, we need to disable tests regularly to avoid circular dependencies caused by test dependencies. 

Running tests in a separate derivation would:
- allow us to test all packages because circular dependencies are no longer a issue;
- reduce build time of the actual packages because tests are ran outside of them;
- add a bit of overhead because there's now two derivations per package;
- allow actual integration testing, because we would build a new env in which we perform the testing.

I propose we then add a `tested` attribute to Python packages that corresponds to the second derivation. This is typically the attribute to use for CI as well.

**Comment by copumpkin** [comment]
I've often wanted this for other types of derivations, too, but I'd definitely support this.

**Comment by anderspapitto** [comment]
One potential snag - there are probably packages where some tests rely on access to temporary build files. 
In those cases, you'd either have to disable any such tests, or alternatively the `tested` attribute could redo the whole build.

**Comment by peterhoeg** [comment]
Earlier conversation #9281.

**Comment by deliciouslytyped** [comment]
It's also very annoying during packaging / nix package development when a test fails after an hour long build and you have to go through the whole process again.

**Comment by doronbehar** [comment]
I don't think we should (by default) add such a `tested` attribute, or put the tests in a `passthru.tests` derivation. In fact, I think that building the tests in the same derivation has the advantage that it forces us to observe the failure, which in turn can warn us about failures also in dependent packages. The same reasoning also pushes me to advocate in transitioning from adding a `passthru.tests.version` to `versionCheckHook`. See also:

- https://nixos.org/manual/nixpkgs/unstable/#tester-testVersion
- https://github.com/NixOS/nixpkgs/pull/320266
- And the discussion that began here: https://discourse.nixos.org/t/improve-the-jdk-infrastructure-on-nixpkgs/45678/14?u=doronbehar .

**Comment by doronbehar** [comment]
If our CI was capable of completely disallowing changing Python derivations (or derivations in general) in a way that makes their `passthru.tests` fail, I wouldn't mind too much separating the tests to their own derivation, and even then, the tooling would have to be _very_ good in order to compensate for the lack of potential tests failure visibility. Our CI and tooling at the moment is very far away from that, and I think that we should first discuss the ideal CI & tooling state we would like to reach, and implement that, before discussing transitioning an ecosystem-specific tests to a separate derivation.

**Comment by yajo** [comment]
`buildPythonPackage` could just split in 2 derivations internally, and always return the tested one, unless `doCheck = false`:

```mermaid
flowchart TD
    buildPythonPackage --> build-system --> wheel
    buildPythonPackage --> dep[dependencies, checkInputs, ...] --> tests
    wheel --> tests
    tests -->|the output is just a symlink to wheel.outPath if tests pass| out

```

**Comment by RuRo** [comment]
FYI, @yajo, if I am understanding your proposal correctly, I think that I've outlined a slightly more complicated version of what you are proposing (with a separate build step and with potentially slightly better parallel build/test support) in [this comment](https://github.com/NixOS/nixpkgs/issues/272178#issuecomment-2575845951).

**Comment by yajo** [comment]
Correct! I didn't read that one before, sorry.

But yes, the basic common idea is to avoid rebuilding when build inputs don't change, but still keep re-running tests (only) when runtime or test inputs change.
